lib/job/arUpdatePublicationStatusJob.class.php

```diff
@@ -18,39 +18,129 @@
+            $this->error($this->i18n->__('Invalid publication status id: %1', ['%1' => $parameters['publicationStatusId']]));
+
+            return false;
+        }
+
         if (null === $resource = QubitInformationObject::getById($parameters['objectId'])) {
             $this->error($this->i18n->__('Invalid description id: %1', ['%1' => $parameters['objectId']]));
 
             return false;
         }
 
-        if (null === $publicationStatus = QubitTerm::getById($parameters['publicationStatusId'])) {
-            $this->error($this->i18n->__('Invalid publication status id: %1', ['%1' => $parameters['publicationStatusId']]));
+        $sql = 'UPDATE status
+            SET status.status_id = :publicationStatus
+            WHERE status.type_id = :publicationStatusTypes ';
 
-            return false;
+        if ($parameters['updateDescendants']) {
+            $sql .= 'AND status.object_id IN (
+                SELECT des.id
+                FROM information_object io
+                JOIN information_object des
+                    ON des.lft >= io.lft AND des.lft < io.rgt
+                JOIN slug
+                    ON io.id = slug.object_id
+                WHERE slug.slug = :slug
+            )';
+
+            $message = $this->i18n->__(
+                'Updating publication status for "%1" and descendants to "%2".',
+                ['%1' => $resource->getTitle(['cultureFallback' => true]), '%2' => $publicationStatus->name]
+            );
+        } else {
+            $sql .= 'AND status.object_id IN (
+                SELECT io.id
+                FROM information_object io
+                JOIN slug
+                    ON io.id = slug.object_id
+                WHERE slug.slug = :slug
+            )';
+
+            $message = $this->i18n->__(
+                'Updating publication status for "%1" to "%2".',
+                ['%1' => $resource->getTitle(['cultureFallback' => true]), '%2' => $publicationStatus->name]
+            );
         }
 
-        $message = $this->i18n->__('Updating publication status for the descendants of "%1" to "%2".', ['%1' => $resource->getTitle(['cultureFallback' => true]), '%2' => $publicationStatus->name]);
         $this->job->addNoteText($message);
         $this->info($message);
 
-        $descriptionsUpdated = 0;
-        foreach ($resource->descendants as $descendant) {
-            $descendant->setPublicationStatus($publicationStatus->id);
-            $descendant->save();
+        $params = [
+            ':publicationStatus' => $publicationStatus->id,
+            ':publicationStatusTypes' => QubitTerm::STATUS_TYPE_PUBLICATION_ID,
+            ':slug' => $resource->slug,
+        ];
+
+        $descriptionsUpdated = QubitPdo::modify(
+            $sql,
+            $params
+        );
+
+        $response = QubitSearch::getInstance()->updateByQuery(
+            [
+                'ancestors' => $resource->id,
+                '_id' => $resource->id,
```

Comment:
Similar here, keeping the resource update outside will remove this clause.

Reply:
👍